### PR TITLE
Patch progress bar

### PR DESCRIPTION
**Pull request by WenzDaniel**
With this pull request we add a nice progessbar. Implemented in get_iter. It also works if the user specifies any time range.

example

### CODE REVIEW

**Comment by jorana** [comment]
Hi Daniel, would it be an option to make the progress-bar optional? Like e.g. https://github.com/JelleAalbers/wimprates/blob/master/wimprates/utils.py#L44.

**Comment by WenzDaniel** [comment]
Hi Joran, thanks. I have to check.

**Comment by WenzDaniel** [comment]
> 
> 
> Hi Daniel, would it be an option to make the progress-bar optional? Like e.g. https://github.com/JelleAalbers/wimprates/blob/master/wimprates/utils.py#L44.

Done, the bar is now optional. By default the bar is switched on. Also added a check that the progress bar is only used if metadata exists.

**Comment by jorana** [comment]
@WenzDaniel some extra questions:
- Do you know how much extra time we need for a typical run to load the data (and does it differ much for event_info / raw_records)? 
- If so perhaps we can make the default different for get_array/get_df might be an option.
- Do you know if it also works for superruns or should we always force it to be false there?

File: strax/context.py [hunk 1]
```diff
@@ -808,19 +818,64 @@ def get_iter(self, run_id: str,
...
+                        time_range=time_range,
+                        time_selection=time_selection)
+                    
+                    if progress_bar:
+                        # Update progressbar:
+                        pbar.n = (result.end - start_time) / (end_time - start_time)
+                        pbar.update(0)
+                        # Now get last time printed and refresh seconds_per_chunk:
+                        # This is a small work around since we do not know the
+                        # pacmaker here.
```
Comment (jorana) [inline strax/context.py:871]: [minor] if you were to do another commit you can correct this to _pacemaker_ (sorry for making these minor comments but I think very few people know what the pacemaker is so it might not ring a bell).

**Comment by WenzDaniel** [comment]
> 
> 
> @WenzDaniel some extra questions:
> 
>     * Do you know how much extra time we need for a typical run to load the data (and does it differ much for event_info / raw_records)?

To be honest I never really checked, since the overall added extra time is super small compared to all the other checks we are doing when reading data. I expect the time difference to be very small. I guess the part which takes longest is the reading in of the metadata for which I have not seen any big differences for the different datakinds.

>     * Do you know if it also works for superruns or should we always force it to be false there?

No, I never worked with superruns and I do not know how they work. If it is a nested loop for run in runid -> for .. in get_iter -> you will get for each run an individual progress bar, if you first somewhat make a "super run" for which  you loop just once with get_iter you will get just a single bar.

File: strax/context.py [hunk 2]
```diff
@@ -808,19 +818,64 @@ def get_iter(self, run_id: str,
...
+                # If no selection is specified we have to get the last end_time:
+                start_time = 0 
+                end_time = float('inf')
+                for t in strax.to_str_tuple(targets):
+                    try:
+                        # Sometimes some metadata might be missing e.g. during tests.
+                        chunks = self.get_meta(run_id, t)['chunks']  
+                        start_time = max(start_time, chunks[0]['start'])
+                        end_time = min(end_time, chunks[-1]['end'])
+                    except (strax.DataNotAvailable, KeyError):
```
[the remark this answers is not in the document]
Comment (WenzDaniel) [reply strax/context.py:836]: > What if one of the chucks raises a KeyError, might you still be able to use the start and end_time of another target (as you loop over them)?

I actually had to add it since pytest would fail otherwise. 

> If so, you might want to set start_time, end_time = None, None above and check later that they aren't None anymore. If they are still None you can then set the progress_bar to False. 

Yes we can do that, but the progress bar may then be a bit odd. In those cases it could stop before 100% or go even further than 100%. Not sure if this would lead to any confusions.

File: strax/context.py (hunk 1 above)
Comment (WenzDaniel) [reply strax/context.py:871]: Totally fine.

File: strax/context.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (WenzDaniel) [reply strax/context.py:836]: Okay I changed it such that if a metadata file of any target exists we will use that one.

**Comment by jorana** [comment]
@ ACs this is ready for merge unless @WenzDaniel feels like addressing CodeFactor

**Comment by jorana** [comment]
@WenzDaniel , this PR would be great but I cannot circumvent the codefactor
